client/src/app/core/core-services/stream.ts

```diff
@@ -47,23 +47,31 @@ export class Stream<T> {
         private messageHandler: (message: T) => void,
         private errorHandler: (error: any) => void
     ) {
-        this.subscription = observable.subscribe((event: HttpEvent<string>) => {
-            if (this.closed) {
-                console.log('got message, but closed');
-                return;
-            }
-            if (event.type === HEADER_EVENT_TYPE) {
-                const headerResponse = event as HttpHeaderResponse;
-                this._statuscode = headerResponse.status;
-                if (headerResponse.status >= 400) {
-                    this.hasError = true;
+        this.subscription = observable.subscribe(
+            (event: HttpEvent<string>) => {
+                if (this.closed) {
+                    console.log('got message, but closed');
+                    return;
+                }
+                if (event.type === HEADER_EVENT_TYPE) {
+                    const headerResponse = event as HttpHeaderResponse;
+                    this._statuscode = headerResponse.status;
+                    if (headerResponse.status >= 400) {
+                        this.hasError = true;
+                    }
+                } else if ((event as HttpEvent<string>).type === PROGRESS_EVENT_TYPE) {
+                    this.handleMessage(event as HttpDownloadProgressEvent);
+                } else if ((event as HttpEvent<string>).type === FINISH_EVENT_TYPE) {
+                    this.errorHandler('The stream was closed');
```

Comment:
Please have a look into OS3's improved error handling: `handleError` accepts more parameters. Have a look into the path the error goes and see where the variables are used. I think especially the server/client error type is used to determine whether to retry or not. Also small thinks like: On the first retry, the timeout is fixed to 500ms.